Fix for issue - #5587

dannyKBjj opened the pull request:
#### Pull Request (PR) description
MSFT_IntuneMobileAppConfigurationPolicyIOS and MSFT_IntuneAndroidManagedStoreAppConfiguration
were supposed to be pulled in favour of existing DSCResource - IntuneAppConfigurationDevicePolicy as my modules duplicated functionality. Whilst testing on our tenant we discovered the following issue: IntuneAppConfigurationDevicePolicy cannot handle 'for Intune' applications #5587

It seems that for some reason only MSFT_IntuneMobileAppConfigurationPolicyIOS was removed. This pull request adds that module back in and deletes "IntuneAppConfigurationDevicePolicy" as I believe that module is now defunct.

See 
https://github.com/microsoft/Microsoft365DSC/pull/5401
https://github.com/microsoft/Microsoft365DSC/issues/5587

#### This Pull Request (PR) fixes the following issues
- Fixes #5587



#### Task list

<!--
    To aid community reviewers in reviewing and merging your PR, please take the time to run
    through the below checklist and make sure your PR has everything updated as required.

    Change to [x] for each task in the task list that applies to your PR. For those task that
    don't apply to you PR, leave those as is.
-->

- [x] Added an entry to the change log under the Unreleased section of the file CHANGELOG.md.
      Entry should say what was changed and how that affects users (if applicable), and
      reference the issue being resolved (if applicable).
- [x] Resource parameter descriptions added/updated in the schema.mof.
- [x] Resource documentation added/updated in README.md.
- [x] Resource settings.json file contains all required permissions.
- [x] Examples appropriately added/updated.
- [x] Unit tests added/updated.
- [x] New/changed code adheres to [DSC Community Style Guidelines](https://dsccommunity.org/styleguidelines).

dannyKBjj committed:
Fix for issue - #5587

MSFT_IntuneMobileAppConfigurationPolicyIOS and
MSFT_IntuneAndroidManagedStoreAppConfiguration
were supposed to be pulled in favour of existing DSCResource - IntuneAppConfigurationDevicePolicy as my modules duplicated functionality. Whilst testing on our tenant we discovered the following issue:
IntuneAppConfigurationDevicePolicy cannot handle 'for Intune' applications #5587

It seems that for some reason only MSFT_IntuneMobileAppConfigurationPolicyIOS was removed. This pull request adds that module back in and deletes "IntuneAppConfigurationDevicePolicy" as I believe that module is now defunct.

dannyKBjj committed:
Added change log updates

FabienTschanz commented:
@dannyKBjj Removing a resource is a breaking change. The next braking change is (I believe) in April 2025. It would be better to add the functionality to `IntuneAppConfigurationDevicePoliy` instead of replacing it with new resources.

dannyKBjj commented:
Hi, but the problem is that it is using convoluted logic to handle both. It is inconsistent with almost every other module (which all handle each workload separately) and has caused an issue, who is to say there aren't other issues? Is it possible to just submit my module alongside it and remove the bugged module at a later date? I don't really want to re-write the existing module as I believe the approach I've taken is better anyway (besides which, I have a working solution already).

FabienTschanz commented:
@dannyKBjj Normally, if there is an issue with a module, we fix the issue in that. The name of the module comes from the separation between the two categories "Managed App" and "Managed Devices". These are also the two things we can select from in the Intune portal, which aligns more with the user experience someone would expect. We already had some cases where we didn't select the "correct" name by workload or by the resource type of what Graph would tell us but rather stay in the Intune portal realms with their naming. 

It's normal for new features and functionality to appear. This happens all the time, so we have to adapt the existing resources for their new purpose. I'd wait until @ykuijs and @NikCharlebois are back and get their opinion as well. I'm just a public contributor adding my input. If you think your solution is better, we'll wait for their feedback. That's totally fine by me.

dannyKBjj commented:
Hi, fair enough... 

I don't think this is a new feature thing, this is just a fault in the way approach taken.

The issue is that the policy pulled by Get-MgBetaDeviceAppManagementMobileAppConfiguration uses the 'ID' of the targeted mobile app in the 'TargetedMobileApps' property, which is a unique value. 

For reasons I don't quite understand existing IntuneAppConfigurationDevicePolicy doesn't use this ID directly, but instead uses it to retrieve the packageId/bundleId. 

The problem is that for these apps iOS and Android will have share the same value, see below:

![image](https://github.com/user-attachments/assets/f278086f-3ed0-40e7-968d-62dd2a480feb)

It then uses value this to retrieve the 'Id' in Set-Target (see below)
![image](https://github.com/user-attachments/assets/cdc10973-ab2d-40ec-9343-6fcb729396f7)

However, because the app/bundleId for iOS and Android is the same, it will simply retrieve both IDs and write them into the policy (which breaks it).

I'll see what is decided in the new year, but from my perspective, the general approach seems to be to write a module for each resource, as DSC doesn't store the platform information. In this case the module seems to be determining it by the presence of certain properties after the fact. Just seems unnecessarily complicated and out of step with most of the other modules.

I can fix the existing one, but seems fixing it is very much the hard route, when I've got 2 modules that already work and were previously accepted to the project? Especially as I feel my approach is more simple, easier to maintain and more consistent with the other DSC resources?

Anyway, have a great Xmas! I'll pick this up when I'm back in the New Year.

dannyKBjj commented:
Any further thoughts on this? Do I need to fix the existing module if I want this PR in the near-term release? To be honest if so, I may just run my modules in my environment and wait for the major update in April?

As I say, I believe the resource type approach is more consistent, easier to maintain etc. Also I don't think this is a new feature problem, I think it is a consequence of the approach taken with the previous module? If my PR will be approved I'll wait, if my PR won't be merged, then I'll reconsider...

Ideally though, it'd be included in a near-term release :-D

ykuijs commented:
@dannyKBjj, thanks for submitting this PR and @FabienTschanz thanks for reviewing it.

Fabien is correct: We only release breaking changes (like removal of resources or parameters) twice a year: First releases of April or October.

In case there is an urgent issue that really breaks something, like the removal of a parameter in a cmdlet that is used, we mark the parameter as Deprecated and leave the parameter. The parameter doesn't work anymore and we write warning messages warning the users that it is deprecated. In the first Breaking change release, the parameter is then removed.

Just to summarize:
- The IntuneMobileAppConfigurationPolicyIOS and IntuneAndroidManagedStoreAppConfiguration resource were added a while back. IntuneMobileAppConfigurationPolicyIOS was removed again, but the IntuneAndroidManagedStoreAppConfiguration resource wasn't (my bad). This was because IntuneAppConfigurationDevicePolicy was already able to configure both and therefore both resource were duplicating functionality. This has been discussed in #5459.
- Now that resource has a different issue. And instead of fixing it, you want to reintroduce the two resources and remove the generic one.
Am I correct?

Just wondering:
- Why isn't fixing the IntuneAppConfigurationDevicePolicy resource possible? If the current logic is incorrect, can be simplified or needs improvement, this would be the prefered option.
- If we indeed decide to move forward, we should not delete the old resources but instead make them deprecated and output warnings. Get and Set methods should output warnings and do nothing. Test method should output warning and always return False.

@ricmestre, what is your opinion on this PR?

ricmestre commented:
@ykuijs My take is still the same when I said to remove the duplicated resource, it's probably better to have separate resources for this and remove the current one but as you've mentioned that would be a breaking change, so yes either the current one must be accommodated to incorporate the fixes or if the new resources are added then the current must have those warnings added.

To be honest in my solution I have the export running automatically so if I update M365DSC I'll get whatever fixes this issue so I don't mind, but for others might not be a solution to have a breaking change outside its schedule.

ykuijs commented:
Ok, then lets move forward with the two resources.

Does the Android version of the resource work correctly right now or does it also need updating?

@dannyKBjj: Can you please update this PR to make the IntuneAppConfigurationDevicePolicy deprecated instead of removing it. The code should provide warnings that the resource is deprecated and do nothing else. The Readme of the resource should also say it is deprecated.

dannyKBjj commented:
Hi yes, I can certainly do that.

I'm not sure what you mean by:

"The code should provide warnings that the resource is deprecated and do nothing else. The Readme of the resource should also say it is deprecated."

Where do I put this code, do you have an example I can refer to?

dannyKBjj committed:
Added module back in, deprecated it as requested/

ykuijs commented:
> Hi yes, I can certainly do that.
> 
> I'm not sure what you mean by:
> 
> "The code should provide warnings that the resource is deprecated and do nothing else. The Readme of the resource should also say it is deprecated."
> 
> Where do I put this code, do you have an example I can refer to?
> 
> Do you mean replace/comment out the code in the old module for Get- Set etc. etc. with a Write-Verbose message to say the module is depracated? Should the message mention the new modules?

Correct, the code of the Get, Set, Test and Export methods should be replaced with a Write-Verbose message stating that the module is deprecated. The same should be added to the Readme.md of the module.

Unfortunately I do not have an example, because we do not have any other resource that is deprecated.

ykuijs commented:
Please also update from Dev. A couple of files have changed

ykuijs commented on Modules/Microsoft365DSC/DSCResources/MSFT_IntuneAppConfigurationDevicePolicy/MSFT_IntuneAppConfigurationDevicePolicy.psm1:
```diff
@@ -95,6 +95,8 @@ function Get-TargetResource
         $AccessTokens
     )
 
+    Write-Verbose -Message "This module has been deprecated, please use MSFT_IntuneMobileAppConfigurationPolicyIOS for iOS policy and MSFT_IntuneAndroidManagedStoreAppConfiguration for Android."
```
This is it. Just add a `return @{}` on the next line to make sure the code stops after this line (same for the other methods).

dannyKBjj replied:
ok, no probs, have done as requested.

dannyKBjj replied:
"Please also update from Dev. A couple of files have changed" I may have messed this up? I fetched Origin in Github desktop, but now I see a conflict?!

dannyKBjj committed:
Added result @{} as requested.

dannyKBjj commented:
Hi, is anyone reviewing this?

ykuijs commented:
> "Please also update from Dev. A couple of files have changed" I may have messed this up? I fetched Origin in Github desktop, but now I see a conflict?!

That is correct. Recently a lot of Intune resources were updated (same issue as described in [here](https://github.com/microsoft/Microsoft365DSC/issues/5640#issuecomment-2604670952)). So please make sure the conflicts are corrected.

ykuijs commented:
@dannyKBjj, can you please update with Dev?

dannyKBjj committed:
Merge remote-tracking branch 'upstream/Dev' into IntuneAppConfigurationDevicePolicy-cannot-handle-'for-Intune'-applications-#5587

dannyKBjj committed:
Updated from DEV to resolve conflicts

As requested

dannyKBjj commented:
Done. Hopefully all OK this time?

dannyKBjj commented:
Now unit tests are failing!!? Again, everything is fine at my end.

dannyKBjj commented:
Same issues with #5588 and #5640
In all cases the the unit test scripts pass in my local environment.
In all cases the test/set/get/export functions are working in my live environment.

ykuijs commented:
I will download your code and run the tests locally to see what is going on.

Some background info:
I am not sure what is going on here. Most of the times these kind of issues are because there are modules present on your system that are not there on the build agent......which means the tests are failing.

What I sometimes do is deleting all modules (except the default ones) from my system or using a freshly installed VM to run the tests.

ykuijs commented:
These tests are failing because the resource has been deprecated and the functions always return an empty hashtable. You should update the unit test for IntuneAppConfigurationDevicePolicy to check for an empty array in each function....that is it. The rest of the tests can be removed.

dannyKBjj commented:
Oh wow, thanks. I'll look into it tomorrow, I could've sworn the errors were in my new modules, but maybe I was hallucinating 😂. Thanks for taking the time to look into it.

dannyKBjj committed:
Fixed test script to pass even though module is deprecated.

suggested to changes script to script didn't seem to work, but what I've done instead allows it to pass, but leaves original code intact (but commented out). Edits should be obvious. Hopefully this is good enough!

dannyKBjj commented:
tried checking for @{} as returned by the function, that didn't seem to work - strange errors along lines of - expected hashtable, but got hashtable errors. Commented out tests instead and checked for $null where applicable - the script now passes and can easily be reverted should anyone wish to.

dannyKBjj commented:
hi, all checks passed on this pull a couple of weeks ago, is it possible that someone could review and (hopefully) merge?

dannyKBjj commented:
hi, still need a review on this please?

dannyKBjj committed:
Preemptively updating Export-DSCResource

recent changes to Export not reflected in this PR. Updating it so expedite module acceptance.

dannyKBjj commented:
Has issue #5588 been fixed somehow? I was looking at simply fixing the module as perhaps that is an easier merge, but now I am going through my steps to reproduce the issue and it seems to be working properly. I checked the code and the logic looks good. I can't find another pull request related to this though.

dannyKBjj commented:
It looks to me like I may as well close this PR?

dannyKBjj commented:
Turns out the issue might've been that you can add the same app twice in the Apps blade. As the original module goes off the name it will pull both bundleIDs. My module didn't have the issue as it goes directly off the bundleID. All in all, it seems like a bug with the Apps blade and not the module, so I'm closing this PR.
